blog/2024-08-07-introducing-secrets-how-wasmcloud-solves-for-secrets-in-distributed-environments/index.mdx

```diff
@@ -0,0 +1,203 @@
+
+* A component or capability provider **requests a secret** by reference to secret's location, without needing to present credential material. 
+* The request goes to a **secrets handler** which in turn requests the specified secret from the appropriate **secrets backend**.
+* Communication with a **secrets store** is mediated by a **secrets backend**.
+
+You can see how a component requests a secret in the diagram below:
+
+![diagram of secret flow when starting component with secret](../../static/img/secrets-diagram.png)
+
+If we zoom into an individual component level, the standard flow for a secret looks like this:
+
+![diagram](./images/component-secret.png)
+
+Let's walk through this process step-by-step:
+
+1. Component sends `wasmcloud:secrets/get` call to host
+2. Host returns secret reference
+3. Component sends `wasmcloud:secrets/reveal` call to host
+4. Host exposes secret from `secrecy` crate's in-memory store, returning value to component
+
+The secret itself is never logged or accessible in memory&mdash;it is encrypted with an [**xkey**](https://docs.nats.io/running-a-nats-service/configuration/securing_nats/auth_callout#encryption), a x25519 keypair, and held in-memory with Rust's [`secrecy` crate](https://crates.io/crates/secrecy). When you drop a secret, it is wiped in-memory. 
+
+## Putting the pieces together
+
+As you can see in the diagram, **secret backends** are an important piece of the puzzle, acting as a mediator between your applications and the secret store(s) of your choice. You can learn more about the design of backends in the documentation, but the most important thing to know up-front is that backends and wasmCloud workloads alike use the common `wasmcloud:secrets` API defined in [WebAssembly Interface Type (WIT)](/docs/concepts/interfaces#webassembly-interface-type-wit). You can [check out the interface on GitHub](https://github.com/wasmCloud/wasmCloud/blob/main/crates/secrets-types/wit/secrets.wit). 
+
+Using a common API enables the same vendor-agnostic approach that we take throughout the rest of the wasmCloud ecosystem&mdash;if you need to swap out your secrets vendor, that's not a problem, as your components and providers remain unchanged while leveraging exactly the same API. All you need to do is change one or two lines of configuration. 
+
+At the launch of secrets support, backends are available for:
+
+* [HashiCorp Vault](https://github.com/wasmCloud/wasmCloud-contrib/tree/main/secrets/secrets-vault)
+* [NATS KV store](https://github.com/wasmCloud/wasmCloud/tree/main/crates/secrets-nats-kv)
+* [Kubernetes Secrets](https://github.com/wasmCloud/wasmCloud-contrib/tree/main/secrets/secrets-kubernetes)
+
+Providing support for Kubernetes Secrets is particularly powerful, since it means that teams [running wasmCloud on Kubernetes](/blog/how-to-run-webassembly-components-on-kubernetes-with-wasmcloud) can immediately use their existing secrets systems with wasmCloud applications. 
```

Comment:
I figured it might be worth illustrating why this is so powerful beyond just the Kubernetes Secrets themselves:
```suggestion
Providing support for Kubernetes Secrets is particularly powerful, since it means that teams [running wasmCloud on Kubernetes](/blog/how-to-run-webassembly-components-on-kubernetes-with-wasmcloud) can immediately use their existing secrets systems with wasmCloud applications. This would even allow you to do things like extending your wasmCloud application secrets to be pulled in via Kubernetes Secrets that are provisioned by an external process such as [External Secrets Operator](https://external-secrets.io/latest/).
```